airflow/providers/amazon/aws/utils/waiter_with_logging.py

```diff
@@ -0,0 +1,77 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import logging
+import time
+
+from botocore.exceptions import WaiterError
+
+from airflow.exceptions import AirflowException
+
+
+def wait(
```

Comment:
It is definitely way cleaner on the operator side! Thanks for doing it. However (and sorry to insist ...), what you are doing here is totally achievable with `tenacity`. You can add logs to every attempt or to every failed attempt, [see documentation here](https://tenacity.readthedocs.io/en/latest/#before-and-after-retry-and-logging). But ... I dont want to hold this PR any longer and I think I expressed myself enough on that issue, at the end of the day it is only question of style :) So LGTM :)

Reply:
Thanks! I tried quite a few different things to get it working with tenacity, but my issue was with debugging. I found that using tenacity made the debugging very difficult. It also added complications in the unit tests (which I was able to resolve, but it took some playing around). 
In hindsight, I think using tenacity to implement a basic retry like this is overkill - tenacity seems to be able to implement fairly complex retry algorithms whereas we were using it as a `while` loop. I would have kept it if things were working smoothly, but I wasn't able to figure out how debugging worked, so I decided to stop before I went further down the rabbit hole :D.

Reply:
I disagree with you, and I don't think the question is whether it's _possible_ to do this with tenacity, but rather if using tenacity improves readability and maintainability.
If I have to read through pages of tenacity documentation to understand what's going on, it's not really a win even if it reduces the number of lines of code.

Tenacity is intended to retry things like network errors & such, using it here as an alternative `while` loop would break the [principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) for readers of this code.